### PR TITLE
Add Kokkos backend

### PR DESCRIPTION
**Pull request by masterleinad**
Builds on top of https://github.com/dealii/dealii/pull/13149 and tries to initialize `Kokkos` when needed and finalize it by registering `Kokkos::finalize` to `std::atexit`. In my tests that worked fine with `Kokkos+Cuda` and initial tests with `Kokkos+Serial` also looked 
good.
It needs more thorough tests, though.

### CODE REVIEW

File: cmake/setup_cached_variables.cmake [hunk 1]
```diff
@@ -370,13 +371,19 @@ UNSET(ENV{NVCCFLAGS})
...
   "If set to ON, then use 64-bit data types to represent global degree of freedom indices. The default is to OFF. You only want to set this to ON if you will solve problems with more than 2^31 (approximately 2 billion) unknowns. If set to ON, you also need to ensure that both Trilinos and/or PETSc support 64-bit indices."
   OFF
   )
 LIST(APPEND DEAL_II_FEATURES 64BIT_INDICES)
 
+CMAKE_DEPENDENT_OPTION(DEAL_II_USE_KOKKOS_BACKEND
+	"Experimental: If set to ON, Kokkos is used for all operations in LinearAlgebra::distributed::Vector. In particular, this allows accessing all Kokkos backends."
+  OFF "DEAL_II_WITH_KOKKOS" OFF
+  )
+LIST(APPEND DEAL_II_FEATURES KOKKOS_BACKEND)
```
Comment (tamiko) [inline cmake/setup_cached_variables.cmake:385]: I know that this is not semantically elegant, but would you mind renaming the variable to `DEAL_II_WITH_KOKKOS_BACKEND`?

The reason being that this entire feature mechanism (i.e., when you add stuff to `DEAL_II_FEATURES` only works with the prefix `DEAL_II_WITH`). For example, a clean configure of your pull request currently shows:
```
#        DEAL_II_WITH_KOKKOS set up with external dependencies
#      ( DEAL_II_WITH_KOKKOS_BACKEND =  )
```

File: include/deal.II/base/config.h.in [hunk 2]
```diff
@@ -46,6 +46,7 @@
 #cmakedefine DEAL_II_WITH_GSL
 #cmakedefine DEAL_II_WITH_HDF5
 #cmakedefine DEAL_II_WITH_KOKKOS
+#cmakedefine DEAL_II_USE_KOKKOS_BACKEND
```
Comment (tamiko) [inline include/deal.II/base/config.h.in:49]: Then this would be not necessary any more :-)

**Review by tamiko — CHANGES_REQUESTED** [review]

**Comment by bangerth** [comment]
I would prefer if the `ensure_kokkos_initialized()` function would simply call `Assert()`. I continue to think that we should initialize Kokkos in `MPI_InitFinalize()` or, alternatively, in a `Kokkos_InitFinalize()` class that one can use in non-MPI programs. If you then have your `ensure_kokkos_initialized()` function simply terminate, this is an indication that a program just needs to add the initialization by hand.

**Comment by masterleinad** [comment]
@bangerth At the moment, I am focussing on getting this pull request working with `Kokkos+CUDA` and `DEAL_II_WITH_CUDA=ON`. Then, I'm curious to see if calling `std::atexit` actually works with that configuration. If it does, I would consider it a long(er) term solution that could be done in a follow-up pull request, though. For now, it's at least helpful to automatically initialize `Kokkos` to be able to run the unit tests without modifying tests.

Of course, I'm open to other suggestions.

**Comment by masterleinad** [comment]
Superseded by #14510.

**Pull request closed without merging**